### PR TITLE
feat(http): add monitorIncomingHTTPRequests config

### PR DESCRIPTION
**Pull request by Qard**
PR for splitting disabling of outgoing http requests from incoming http requests.

Closes elastic/apm-agent-nodejs#434

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update documentation

### CODE REVIEW

**Comment by watson** [comment]
Tip: You can use the new Draft PR feature of GitHub for WIP PR's 😃

File: test/instrumentation/_agent.js [hunk 1]
```diff
@@ -12,27 +13,14 @@ var sharedInstrumentation
...
 module.exports = function mockAgent (expected, cb) {
   var agent = {
-    _conf: {
-      serviceName: 'service-name',
-      active: true,
-      instrument: true,
-      captureSpanStackTraces: true,
+    _conf: config({
+      abortedErrorThreshold: '250ms',
+      asyncHooks: false,
```
Comment (watson) [inline test/instrumentation/_agent.js]: Why did you add this?

File: test/instrumentation/_agent.js [hunk 2]
```diff
@@ -12,27 +13,14 @@ var sharedInstrumentation
...
+      breakdownMetrics: false,
```
Comment (watson) [inline test/instrumentation/_agent.js]: Why did you add this?

File: test/instrumentation/_agent.js (hunk 1 above)
Comment (Qard) [reply test/instrumentation/_agent.js]: There was a test issue...I forget exactly what it broke. Technically though, the _omission_ of `asyncHooks: false` previously made it _implicitly_ false and this just restores the previous behaviour.

File: test/instrumentation/_agent.js (hunk 2 above)
Comment (Qard) [reply test/instrumentation/_agent.js]: Just to be _extra_ sure it's not producing any metrics stuff that'd mess with event count results? 🤷‍♂

File: lib/config.js [hunk 3]
```diff
@@ -69,6 +69,7 @@ var DEFAULTS = {
   logLevel: 'info',
   metricsInterval: '30s',
   metricsLimit: 1000,
+  monitorIncomingHTTPRequests: true,
```
[the remark this answers is not in the document]
Comment (Qard) [reply lib/config.js]: I was actually thinking the same thing that maybe it should use "instrument" instead.

File: test/instrumentation/_agent.js (hunk 2 above)
Comment (watson) [reply test/instrumentation/_agent.js]: But this is just a mock agent, so it shouldn't start any metrics stuff right?

File: test/instrumentation/_agent.js (hunk 1 above)
Comment (watson) [reply test/instrumentation/_agent.js]: As this is a mock agent I'm just wondering where this makes a difference... and since you didn't add any new tests, it's weird that this update should change anything. So I'm a bit worried if it's covering over something else 🤔

File: test/instrumentation/_agent.js (hunk 2 above)
Comment (Qard) [reply test/instrumentation/_agent.js]: No, it starts the metrics stuff. Not sure _why_, if it's being disabled again by `metricsInterval: 0`, but it's been that way for awhile. 🤷‍♂

Comment (watson) [reply test/instrumentation/_agent.js]: I originally thought that since we don't call the `agent.start()` function, then we never start the metrics, but now I just realize that we manually start the metrics in this mock agent:

https://github.com/elastic/apm-agent-nodejs/blob/8aa4344cedfb9515bca21c7efa89da4f4c6e2bcf/test/instrumentation/_agent.js#L34-L35

Do you remember why we do that? I mean, if we in the config completely disable it, then why try to start it?

File: lib/config.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Qard) [reply lib/config.js]: fixed

**Comment by Qard** [comment]
I've narrowed the 12.0 failure to `http.request(...)` losing context before emitting the `'response'` event, which calls the response handler function, if given. I'm not sure yet exactly _where_ that context loss occurs between the the call and the handler. Possibly some internals in `net`? 🤔

**Comment by Qard** [comment]
The failure is due to the issue fixed by https://github.com/nodejs/node/pull/27477, which didn't land until 12.2.0. As a result, all 12.x releases _before_ 12.2.0 will actually have a context loss in the response handler of all outgoing requests. Apparently our existing tests just happened to not cover that particular case. 🤔

File: test/instrumentation/modules/http/disabling.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    function ping (url, fn = res => res.resume()) {
+      return new Promise((resolve, reject) => {
+        const req = http.get(url, res => {
+          res.on('error', reject)
+          res.on('end', resolve)
+          fn(res)
+        })
+        req.on('error', reject)
+        req.end()
```
Comment (watson) [inline test/instrumentation/modules/http/disabling.js]: No need to call `end()` on a `http.get()` request as they are automatically ended

File: lib/instrumentation/http-shared.js [hunk 5]
```diff
@@ -156,6 +156,7 @@ exports.traceOutgoingRequest = function (agent, moduleName, method) {
       return req
 
       function onresponse (res) {
+        ins._recoverTransaction(span.transaction)
```
Comment (watson) [inline lib/instrumentation/http-shared.js]: It would be nice also get this as a separate PR so we can get it into 2.x

**Comment by watson** [comment]
The PR description still says WIP, but I assume it's ready now?

**Comment by Qard** [comment]
Yep, just forgot to update the description. 👍
